### DISCUSSION TITLE
Send input to multiple panes

### DISCUSSION THREAD

**Discussion by diego-carrera** [ghostty-org/ghostty#3227] [Feature Requests, Ideas]
It would be great to have the ability to send the same input to multiple panes in the current active tab.

iTerm does this and is greatly missed.

**Comment by MassimilianoPasquini97** [comment]
For me is the only feature I'm missing from iTerm2. For now I will use tmux when I will need to broadcast input to more panes but it will be awesome if can be implemented as Ghostty Action.

**Comment by YLivay** [comment]
Seconding this.

This feature exists in [Tilix](https://github.com/gnunn1/tilix) and in Kitty using the [broadcast kitten](https://sw.kovidgoyal.net/kitty/kittens/broadcast/)

This is huge time saver for me as a dev ops, i would absolutely love to see it in Ghostty.

**Reply by timelmer-uw** [reply]
Just want to toss in my two cents since you mentioned it. Tilix is great. That Tilix works in Ghostty is great. But since Ghostty supports split panes, I'd also like it to support this feature.

Currently the only blocker for me moving from Konsole.

**Comment by canTuncPekkan** [comment]
can't you do this with the prefix all:

`all:` - Make the keybind apply to all terminal surfaces. By default, keybinds only apply to the focused terminal surface. If this is true, then the keybind will be sent to all terminal surfaces. This only applies to actions that are surface-specific. For actions that are already global (i.e. `quit`), this prefix has no effect.

or are you asking for a more selective version of this.

**Reply by MassimilianoPasquini97** [reply]
> can't you do this with the prefix all:
> 
> `all:` - Make the keybind apply to all terminal surfaces. By default, keybinds only apply to the focused terminal surface. If this is true, then the keybind will be sent to all terminal surfaces. This only applies to actions that are surface-specific. For actions that are already global (i.e. `quit`), this prefix has no effect.
> 
> or are you asking for a more selective version of this.

We are asking the possibility to broadcast input to multiple panes, in tmux term: "setw synchronize-panes on" and "setw synchronize-panes off". On iTerm2 this feature was implemented natively as a keybinds: "CMD+SHIFT+I" for Broadcast input to all panes in all tabs and "CMD+OPTION+I" for Broadcast input to all panes in current tab.

In my workflow this feature is very useful when you need to input the same command across multiple ssh sessions. Some network switches in HA or staked  don't sync config 😅.

This feature implemented natively on Ghostty as Action will be the cherry on top.

**Comment by YLivay** [comment]
I don't know any Zig but this repo is so delightfully readable I couldn't help poking around a bit.

Looks like input is read in a separate thread in Exec.zig. input is sent to termio.Termio.processOutput which ends up calling Terminal.print. Terminal owns a Screen which has a PageList that is essentially a linked list of "pages", which is the individual terminal pane. Look like we'd have to spread the input across these panes.

Issue is, the Screen is the one that owns the concept of a "cursor", not the individual pages. Attached to the cursor are other concepts that would need addressing like knowing which link is currently active (the url under the cursor), the cursor's current style (which the renderer assumes there is only one of), the concept of page "pins", how that interacts with mouse controls (absolute vs relative cursor movements), and in general a bunch of optimizations that rely on there being only one cursor.

I think adding multi cursor support is a big ask, but maybe its easier to split the functionality of typing in a pane to a new type of cursor that is local to the pane while the screen's cursor remains the global "interactive" cursor?

Edit: Actually, it'd probably be easier to hack in something that emulates synchronized inputs by having each pane remember keep a copy of the global cursor at the point in time when the cursor was last active in the pane. then, on input, iterate through all panes, restore the global cursor to the pane's cursor, apply the input and at the end restore to the global cursor?

**Comment by memco** [comment]
> can't you do this with the prefix all:

The docs sound like `all:` only applies to a specific keybind for a single specific action, but the feature request is like multiple cursors or multiple focused surfaces at the same time like @MassimilianoPasquini97  said so that you could enter the same text into multiple different SSH sessions without having to enter the command in one surface, switch the the next and repeat. 

> Edit: Actually, it'd probably be easier to hack in something that emulates synchronized inputs by having each pane remember keep a copy of the global cursor at the point in time when the cursor was last active in the pane. then, on input, iterate through all panes, restore the global cursor to the pane's cursor, apply the input and at the end restore to the global cursor?

I only have iTerm's implementation as reference, but this sounds roughly like what I expect. iTerm seems to internally keep a single active pane as after you disable input broadcast it goes back to whichever pane was last active and the cursor is in the same state it was just in. I found [this iTerm page with a sample python script](https://iterm2.com/python-api/examples/broadcast.html) that might be of interest.

**Comment by sadok-f** [comment]
It is definitely the only feature that I'm missing from iTerm2.
it would be great to have it implemented.

**Reply by Spiral90210** [reply]
Same, it's the only thing I cannot do which I need ocasionally e.g. sweeping dependency updates or go versions in 15+ repos. Just need to have whatever I've typed be sent into all panes in the visible tab. Don't care where the cursor is or anything like that, I can adjust each individually by dropping to single focus if needed. Also not worried about having it go back to the last 'active' pane when deactivated, although that is a nice touch.

**Comment by darvad** [comment]
+1 from my side. 
Really like Ghostty so far. Have it setup in no time to my liking. Very nice out of the box defaults. 
I am a network engineer and often ssh to multiple switches and routers. Each connection in its own pane. Being able to send commands to all panes the same time like with iterm2 or tmux is a must for me. Would really love to see that feature built natively into to ghostty and get rid of additional tmux layer.

**Comment by tantanas** [comment]
+1
absence of broadcast to multiple panes is the only feature i'm missing for my workflow too.
hopefully a straight forward feature to implement.

**Comment by whammad-nfi** [comment]
+1

**Comment by BeKnowDo** [comment]
+1

**Comment by htomasz** [comment]
+1

**Comment by 00-kat** [comment]
Seriously folks?  *Five* `+1`s in a row?  GitHub literally CSS'd a button onto your screen so that you can effectively communicate your interest:
![A screenshot of the upvote button on the top of the post.](https://github.com/user-attachments/assets/51a73fa8-1f1c-45a0-906d-0cc544ac1429)
Everyone subscribed is getting an email every time someone adds a useless `+1` comment.  Please stop it.

And for the record: simply pressing that upvote button bumps a thread to the top too; you don't *need* to add a comment if that was what you were intending to achieve.

**Comment by jcollie** [comment]
Locked to contributors, since y'all aren't listening to @00-kat .
